### PR TITLE
Add HashAlgorithm hash and hashWithTag functions

### PR DESCRIPTION
**Pull request by turbolent**
## Description

- Add two functions to `HashAlgorithm`: `hash` and `hashWithTag`
- Use these new functions in the `Crypto` contract instead of passing in a hasher. This basically decouple the contract from Cadence, so it can now be deployed on-chain
- Add a hash handler to the interpreter, and route it to the runtime interface
- Refactor the existing public key validation and signature verification to always use the interpreter's handlers
- Allow the import of hash algorithms. Use a special constructor to declare the built-in methods
- Test import of signature algorithms and hash algorithms

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/crypto_test.go [hunk 1]
```diff
@@ -30,35 +31,6 @@ import (
 	"github.com/onflow/cadence/runtime/tests/utils"
 )
 
-func TestRuntimeCrypto_import(t *testing.T) {
```
Comment (turbolent) [inline runtime/crypto_test.go:33]: This is a test for old behaviour which is removed now. It also should have just been a checker test

File: runtime/crypto_test.go [hunk 2]
```diff
@@ -253,28 +218,9 @@ func TestRuntimeCrypto_hash(t *testing.T) {
 		assert.True(t, called)
 		assert.Equal(t, "some-tag", hashTag)
 	})
-
-	t.Run("hashWithTag - without tag", func(t *testing.T) {
```
Comment (turbolent) [inline runtime/crypto_test.go:257]: This is / should be a checker test

File: runtime/interface.go [hunk 3]
```diff
@@ -291,13 +291,3 @@ func (i emptyRuntimeInterface) GetStorageCapacity(_ Address) (uint64, error) {
 func (i *emptyRuntimeInterface) ValidatePublicKey(_ *PublicKey) (bool, error) {
 	return false, nil
 }
-
-func (i emptyRuntimeInterface) ValidateSignature(
```
Comment (turbolent) [inline runtime/interface.go:295]: This function was renamed and this implementation was left over

File: runtime/interpreter/interpreter.go [hunk 4]
```diff
@@ -410,12 +418,22 @@ func WithPublicKeyValidationHandler(handler PublicKeyValidationHandlerFunc) Opti
 	}
 }
 
-// WithSignatureValidationHandler returns an interpreter option which sets the given
+// WithSignatureVerificationHandler returns an interpreter option which sets the given
 // function as the function that is used to handle signature validation.
 //
-func WithSignatureValidationHandler(handler SignatureValidationHandlerFunc) Option {
+func WithSignatureVerificationHandler(handler SignatureVerificationHandlerFunc) Option {
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:424]: Make naming consistent public key validation, but signature verification

File: runtime/stdlib/contracts/crypto.cdc [hunk 5]
```diff
@@ -1,21 +1,12 @@
...
-        tag: String,
-        algorithm: HashAlgorithm
-    ): [UInt8]
-}
-
 pub contract Crypto {
 
     pub fun hash(_ data: [UInt8], algorithm: HashAlgorithm): [UInt8] {
-        return self.hashWithTag(data, tag: "", algorithm: algorithm)
+        return algorithm.hash(data)
```
Comment (tarakby) [inline runtime/stdlib/contracts/crypto.cdc:5]: 👌

File: runtime/stdlib/contracts/crypto.cdc [hunk 6]
```diff
@@ -160,14 +151,7 @@ pub contract Crypto {
...
 
-    priv let hasher: {Hasher}
-
-    init(hasher: {Hasher}) {
-
-        self.hasher = hasher
-
-        // Initialize constants
-
+    init() {
```
Comment (tarakby) [inline runtime/stdlib/contracts/crypto.cdc:154]: 😍

File: docs/language/crypto.mdx [hunk 7]
```diff
@@ -31,6 +31,12 @@ pub enum HashAlgorithm: UInt8 {
     /// This is a customized version of KMAC128 that is compatible with the hashing to curve 
     /// used in BLS signatures.
     pub case KMAC128_BLS_BLS12_381 = 5
+
+    /// Returns the hash of the given data
+    pub fun hash(_ data: [UInt8]): [UInt8]
```
Comment (tarakby) [inline docs/language/crypto.mdx:36]: I think we can update the doc by taking the hashing section out of the crypto contract section. 
We could also update the example [here](https://github.com/onflow/cadence/pull/1002/files#diff-7aeb0f4db759cececb1b38ca7d84c09a8c4ddf85b3ef7e5195f1af04f59b689eL229) to not use the crypto contract.

**Review by tarakby — APPROVED** [review]
Nice updates 🏄

**Review by SupunS — APPROVED** [review]
Looks great! and a nice clean up too!

File: docs/language/crypto.mdx (hunk 7 above)
Comment (turbolent) [reply docs/language/crypto.mdx:37]: Good idea! I moved the hashing documentation up to the hash algorithm overview in dbfcacc83f48e86c2b623e06c5a755adc1100285. I also fixed some other things

**Pull request merged**